src/crypto/tls/ech.go

```diff
@@ -0,0 +1,821 @@
+		b.AddBytes(handle.enc)
+	})
+	return b.BytesOrPanic()
+}
+
+// ecgUnmarshalClient parses the prefix of the ClientECH used by the server to
+// setup its HPKE context.
+func echUnmarshalContextHandle(raw []byte) (*echContextHandle, error) {
+	s := cryptobyte.String(raw)
+	handle := new(echContextHandle)
+	if !echReadContextHandle(&s, handle) || !s.Empty() {
+		return nil, fmt.Errorf("error parsing context handle")
+	}
+	handle.raw = raw
+	return handle, nil
+}
+
+func echReadContextHandle(s *cryptobyte.String, handle *echContextHandle) bool {
+	var t cryptobyte.String
+	if !s.ReadUint16(&handle.suite.kdfId) ||
+		!s.ReadUint16(&handle.suite.aeadId) ||
+		!s.ReadUint8LengthPrefixed(&t) ||
+		!t.ReadBytes(&handle.configId, len(t)) ||
+		!s.ReadUint16LengthPrefixed(&t) ||
+		!t.ReadBytes(&handle.enc, len(t)) {
+		return false
+	}
+	return true
+}
+
+// echIncorporateOuterExtensions interprets data as a ClientHelloInner message
+// and transforms it as specified by the "outer_extension" extension. It returns
+// the transformed ClientHelloInner and a bit indicating if parsing succeeded.
+//
+// outerExtensions specifies the sequence of extensions that will be compressed.
+// This function panics if this sequence is non-contiguous. The handshake will
+// fail if the outer extensions do not appear in the ClientHelloInner.
+func echIncorporateOuterExtensions(data, innerDigest []byte, outerExtensions []uint16) ([]byte, bool) {
+	var b cryptobyte.Builder
+	idx, ok := getIndexOfClientHelloExtensions(data)
+	if !ok {
+		return nil, false
+	}
+
+	b.AddBytes(data[:idx])
+	s := cryptobyte.String(data[idx:])
+	if s.Empty() {
+		return b.BytesOrPanic(), true
+	}
+
+	var extensions cryptobyte.String
+	if !s.ReadUint16LengthPrefixed(&extensions) || !s.Empty() {
+		return nil, false
+	}
+
+	var errReadFailure = errors.New("read failure")
+	b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
+		for !extensions.Empty() {
+			var ext uint16
+			var extData cryptobyte.String
+			if !extensions.ReadUint16(&ext) ||
+				!extensions.ReadUint16LengthPrefixed(&extData) {
+				panic(cryptobyte.BuildError{Err: errReadFailure})
+			}
+
+			if ext == outerExtensions[0] {
+				// Replace outer extensions with "outer_extension" extension.
+				b.AddUint16(extensionECHOuterExtensions)
+				b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
+					b.AddUint16LengthPrefixed(func(b *cryptobyte.Builder) {
+						for _, outerExt := range outerExtensions {
+							b.AddUint16(outerExt)
+						}
+					})
+					b.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
+						b.AddBytes(innerDigest)
+					})
+				})
+
+				// Discard the outer extensions.
+				for _, outerExt := range outerExtensions[1:] {
+					if ext != outerExt || extensions.Empty() {
+						panic("internal error: malformed outer extensions")
+					}
+					if !extensions.ReadUint16(&ext) ||
+						!extensions.ReadUint16LengthPrefixed(&extData) {
+						panic(cryptobyte.BuildError{Err: errReadFailure})
+					}
```

Comment:
Since only one KeyShare extension is supported, let's simplify this code until more extensions are actually supported. The comment `Replace outer extensions with "outer_extension" extension.` could then be clarified to something like `Replace "key_share" extension with "outer_extension" extension.`

Reply:
That would break compatibility with clients that compress a sequence of extensions other than ["key_share"].

Reply:
Done.